react/README.md

```diff
@@ -1259,34 +1267,338 @@ For more info on `dangerouslySetInnerHTML`, see: [Dangerously Set innerHTML](htt
+    }
+
+    return (
+        <div>
+            {header}
+            {buttons}
+        </div>
+    );
+}
+
+// bad (expressions in JSX)
+render() {
+    let {includeHeader} = this.props;
+
+    return (
+        <div>
+            {includeHeader ? (<h2>Pagination</h2>) : null}
+            {[1, 2, 3, 4, 5].map((page) => (
+                <Button key={page} onClick={this._handlePageClick.bind(this, page)} />
+            ))}
+        </div>
+    );
+}
+```
+
+The above "bad" example doesn't seem so bad right? But as we know, code tends to grow over time. If we add more expressions, add more markup to the header, or the map gets more more logic, the code will become unwieldy. Setting up this guideline, even in the most simple examples, helps set the code along the right path.
+
+See [Helper components](#helper-components) for another way to help keep `render()` lean.
+
+**[⬆ back to top](#table-of-contents)**
+
+### Hiding elements
+
+With React's optimized re-rendering via its Virtual DOM abstraction, you should never need to hide elements with CSS (except maybe with some sophisticated CSS animations). Instead, don't render the element when it shouldn't be visible, and render it when it should:
+
+```js
+// good
+export default class Togglr extends React.Component {
+    state = {visible: false}
+
+    _handleToggle() {
+        this.setState({
+            visible: !this.state.visible
+        })
+    }
+
+    render() {
+        let {visible} = this.state;
+        let message;
+
+        if (visible) {
+            message = (
+                <p>This message is toggled on/off with React not CSS!</p>
+            );
+        }
+
+        return (
+            <div>
+                <Button click={this._handleToggle.bind(this)}>Toggle!</Button>
+                {message}
+            </div>
+        );
+    }
+}
+
+// bad (uses CSS to hide element instead of not rendering)
+export default class Togglr extends React.Component {
+    state = {visible: false}
+
+    _handleToggle() {
+        this.setState({
+            visible: !this.state.visible
+        })
+    }
+
+    render() {
+        let {visible} = this.state;
+        let messageClassName;
+
+        if (!visible) {
+            messageClassName = 'hidden';
+        }
+
+        return (
+            <div>
+                <Button click={this._handleToggle.bind(this)}>Toggle!</Button>
+                <p className={messageClassName}>
+                    This message is toggled on/off with CSS 🙁!
+                </p>
+            </div>
+        );
+    }
+}
+```
+
+_Friendly reminder:_ If you want an **entire** component to be conditionally rendered, the component must return `null`. Returning `undefined` will be an error.
 
 **[⬆ back to top](#table-of-contents)**
 
 ## State
 
+There are two ways of maintaining data in a React component: props and state.
+
+The props are used by a component's parent to configure the component and are immutable within the component.
```

Comment:
nitpick: Might be clearer visually to separate the props and state descriptions with small headers or at least bolding. ie **Props**: "asdfasdf" **State**:"sfgdgfsdfg"